### PR TITLE
Fix `AccessRequest` permissions in user ACL

### LINKED ISSUE #20135: Access Requests not displayed in UI with the default 'access' role
Expected behavior:

The Access Requests section should be displayed in the Web UI with the default `access` role.

Current behavior:

Now that [Issue 19054](https://github.com/gravitational/teleport/issues/19054) has been implemented, the Access Requests section is not displayed in the Web UI with the default `access` role.

In order to view the Access Requests section, the default `access` role needs to include the following permissions:

```
    - resources:
      - access_request
      verbs:
      - list
      - create
      - read
```

Bug details:
- Teleport version: 11.2.1
- Recreation steps: Deploy a new Teleport cluster, create a user with the default `access` role and access the web UI.
- Debug logs: N/A

### PR DESCRIPTION
**Pull request by mcbattirola**
This PR fixes https://github.com/gravitational/teleport/issues/20135

Since https://github.com/gravitational/webapps.e/pull/461 was added, users without `create` permission on `access_request` can't see the Create Access Request screen in the web UI.

This is a bug, since `create` on `access_request` is not necessary for a user to create an access request. Instead, as long as they have a requestable role or a `search_as_role`, they are able to create an access request.

This PR fixes that by implementing the logic described above.

It also addes read and write permissions for `access_request` to the default role `access`, so new users in a new cluster will see the screen. They won't be able to actually make access requests, since they don't have `search_as_roles`, but this is the same behavior we had before  https://github.com/gravitational/webapps.e/pull/461.

Closes https://github.com/gravitational/teleport/issues/20135

### CODE REVIEW

File: lib/web/ui/usercontext.go
Comment (zmb3) [inline lib/web/ui/usercontext.go:223]: Is `Edit` needed, or is that only for approvals?

What about delete?

File: lib/web/ui/usercontext.go
Comment (rudream) [inline lib/web/ui/usercontext.go:221]: They should only be granted access to `list`, `read` and `create`. They shouldn't be able to have the ability to `edit` (approve/deny) and `remove` just as a result of having requestable roles defined

Comment (zmb3) [reply lib/web/ui/usercontext.go:221]: What if you want to add a resource to the request before it gets approved? You'd need edit for that, right?

Comment (mcbattirola) [reply lib/web/ui/usercontext.go]: Thanks, fixed in https://github.com/gravitational/teleport/pull/20209/commits/877f6127cdbe23d52bca52d8f712e6d96161946b

Also made the `access` role only have `list`, `read` and `create`.

File: lib/web/ui/usercontext.go
Comment (mcbattirola) [reply lib/web/ui/usercontext.go]: Yes, this should be only `read`, `create` and `list`. Fixed in https://github.com/gravitational/teleport/pull/20209/commits/877f6127cdbe23d52bca52d8f712e6d96161946b

File: lib/web/ui/usercontext.go
Comment (mcbattirola) [reply lib/web/ui/usercontext.go]: > What if you want to add a resource to the request before it gets approved? You'd need edit for that, right?

Yes. Also, you can't delete your own access request through the web UI if we set `Delete` to false here.

I think we can return all to true here for the web UI sake, and if they try to delete or edit an acess request that is not theirs, they will get a permission denied error

**Comment by rudream** [comment]
We should probably also update the documentation for role configs to make it clear that having requestable roles or `search_as_role`'s defined will automatically grant the ability to `list`, `read`, `create` and `edit` Access Requests, even if that access isn't explicitly granted in the role.

File: lib/services/presets.go
Comment (nklaassen) [inline lib/services/presets.go:124]: wait, why do we need to add new rules for this? Seems like a front end/display issue. Will this not allow users with `access` to see requests created by other users and/or create requests for other users through the API?

File: lib/web/ui/usercontext.go
Comment (nklaassen) [inline lib/web/ui/usercontext.go:217]: what about reviewers who cannot themselves create access requests, they should also be able to see the page

File: lib/services/presets.go
Comment (mcbattirola) [reply lib/services/presets.go:124]: Yes, it will.

What about adding just `create` here and sending all permissions as `true` to the frontend if the user have any permission at all?

**Comment by mcbattirola** [comment]
> We should probably also update the documentation for role configs to make it clear that having requestable roles or search_as_role's defined will automatically grant the ability to list, read, create and edit Access Requests, even if that access isn't explicitly granted in the role.

True, I'll updated it.

Actually, only creating and listing your own access requests is enabled without permission. Delete is unauthorized for your own access requests if you don't have `delete` permission, not sure if intended or a bug.

**Review by nklaassen — CHANGES_REQUESTED** [review]
ACLs are the wrong way to think about controls on access requests. Submitting an access request is not the same as creating an access requests resource, submitting a review is not the same as editing an access request resource.

I would double check this, but the ACL model treats the access request as just another backend resource. If a role has a `create` ACL for access requests it is blanket permission to create an access request for any user in any state - we don't want that.

See `CreateAccessRequest` in ServerWithRoles, if they have the ACL they can do anything. I don't think our docs ever recommend setting those.
https://github.com/gravitational/teleport/blob/6c46eac76a8fce7b13fec4b11d077bca46a0ebdc/lib/auth/auth_with_roles.go#L1974-L1982

File: lib/services/presets.go
Comment (nklaassen) [reply lib/services/presets.go:124]: we also should not have `create` here. One should just never use ACLs with access requests (unless you are writing a bot or debugging with tctl as a super-admin or something)

Comment (nklaassen) [reply lib/services/presets.go:124]: FYI it doesn't really matter to me what the fix is for the web UI, I have requested changes on this PR just because I strongly believe we should not be adding any new access request ACLs to any roles

Comment (mcbattirola) [reply lib/services/presets.go:124]: I see, thanks for the explanation about ACLs and Access Requests, that makes sense. I had wrong assumptions there.

**Comment by mcbattirola** [comment]
Ok, given @nklaassen's explanation, I see two possible ways forward:

1. Always show the Access Requests screens in the web UI unless the user has an explicit `deny` rule on access requests. Again, this is not how you're supposed to use Access Requests, so not great.

2. Only show Access Requests screens in the web UI if a user's role have `spec.allow.search_as_roles` or `spec.allow.review_requests`. In this case, `access` won't see the screen by default.

File: lib/web/ui/usercontext.go
Comment (mcbattirola) [reply lib/web/ui/usercontext.go:217]: Thanks, added the following block to allow reviewers to see the screens:
```go
	// If the user has preview_requests or preview_as_roles,
	// they should be able to see the Access Requests screens in order to review requests
	if len(userRoles.GetAllowedPreviewAsRoles()) > 0 || userRoles.WithoutImplicit().MaybeCanReviewRequests() {
		requestAccess = access{
			List:   true,
			Read:   true,
			Create: true,
			Edit:   true,
			Delete: true,
		}
	}
```

**Comment by mcbattirola** [comment]
Closing this PR for now, I'll revert the changes in webapps.e (revert PR: https://github.com/gravitational/webapps.e/pull/473/files).

**Comment by rudream** [comment]
@mcbattirola Another approach that could resolve all the issues without changing the ACL could be to have all the requestable roles and `search_as_roles` included in the `UserContext` sent to the Web UI (there already is a `requestableRoles` field in there), then the UI could render the Access Requests tab on a condition like this:

```
if(userAcl.accessRequests.list || userContext.requestableRoles?.length || userContext.searchAsRoles?.length)
```

Thoughts?

**Comment by mcbattirola** [comment]
@rudream yeah, I think this make a lot of sense.

Only problem is `access` won't see the Access Requests screens. If we're OK with this, I think this is the best solution

**Comment by nklaassen** [comment]
> Only problem is `access` won't see the Access Requests screens. If we're OK with this, I think this is the best solution

I don't think `access` should be able to see it, it does not have permission to create or review any requests

**Comment by ravicious** [comment]
I think we should coordinate this change with Connect as well.

At the moment, Connect pings the auth server and looks at `PingResponse.ServerFeatures.AdvancedAccessWorkflows` to decide whether to show the UI for access requests. If the cluster supports access requests but the user doesn't have permission to interact with them, the UI will return an error message when attempting to perform certain actions, similar to how the Web UI works right now.

**Pull request closed without merging**
